### PR TITLE
Fix chi2 and errors in fit function

### PR DESCRIPTION
**Pull request by bpalmeiro**
- Chi2
Implemented chi2 computation, based on scipy.stat.chisquare is only valid when data smearing is Poisson distributed. So, a general chi2 function is added, with formula extracted from [1].

- Errors
Fit errors, given by scipy.optimize.curve_fit were not correct as flag absolute_sigma was set to false. When true, "[data] sigma describes one standard deviation errors of the input data points", so covariance matrix estimation is based on these values. Otherwise, data sigma is only taken as "relative weights of the data points" and covariance matrix is not affected by the overall magnitude of the values in sigma, just by relative magnitudes.

Both improvements are shown in https://github.com/bpalmeiro/ICARO/blob/Chi2_demostration/icaro/Chi2_demonstration/Chi2_demonstration.ipynb

[1]  Numerical Recipes. The Art of Scientific Computing, 3rd Edition, 2007, ISBN 0-521-88068-8. Pages 778-780

### CODE REVIEW

File: invisible_cities/core/fit_functions.py [hunk 1]
```diff
@@ -27,6 +27,35 @@ def get_errors(cov):
...
+
+    Parameters
+    ----------
+    ydata : np.ndarray
+        Data array
+    ydata : np.ndarray
+        Fit funtion values corresponding to ydata array
+    sigma : np.ndarray
+        Data errors array. If no sigma is given, it takes the poisson case.
+
```
Comment (gonzaponte) [inline invisible_cities/core/fit_functions.py]: ndf is not documented

File: invisible_cities/core/fit_functions.py [hunk 2]
```diff
@@ -96,16 +125,26 @@ def fit(func, x, y, seed=(), fit_range=None, **kwargs):
     vals, cov = scipy.optimize.curve_fit(func,
                                          x, y,
                                          seed,
+                                         absolute_sigma=True,
                                          **kwargs)
 
     fitf = lambda x: func(x, *vals)
     fitx = fitf(x)
-    chi2, pval = scipy.stats.chisquare(y[fitx != 0], fitx[fitx != 0], len(vals))
+
```
Comment (gonzaponte) [inline invisible_cities/core/fit_functions.py]: use a variable to hold the result of `fitx != 0`. It is used in 4 different places.

File: invisible_cities/core/fit_functions.py [hunk 3]
```diff
@@ -96,16 +125,26 @@ def fit(func, x, y, seed=(), fit_range=None, **kwargs):
...
+    if not "sigma" in kwargs:
+        sigma_r   = None
```
Comment (gonzaponte) [inline invisible_cities/core/fit_functions.py]: You can use `sigma_r = kwargs.get("sigma")` which returns `None` if the key is not found.  Or `sigma_r = kwargs.get("sigma", None)`, which is more explicit.

**Review by gonzaponte — CHANGES_REQUESTED** [review]
Great work! Very well documented and demonstrated.

**Comment by jjgomezcadenas** [comment]
Really good work. The Galician gang (GG's) are increasing even more their level of Karma.
I suggest to turn the examples in the notebooks into tests! Simply take the results that you obtain for chi2 and P-value with both methods and assert them. Like that the test (with a few comments) also documents the work!

Incidentally, the result:
chi2     : 0.9911404817192085
P-value  : 1.328398e-21

Which you obtain for Poissonian errors with old method does not make sense for P-value. 

Good job (I will live @gonzaponte to approve it, after you add the little changes he suggests and perhaps the docs-test. :)!

File: invisible_cities/core/fit_functions.py (hunk 1 above)
Comment (gonzaponte) [reply invisible_cities/core/fit_functions.py]: I don't want to be pedantic, but can we replace `len(data)` by `number of data points`?

**Review by gonzaponte — DISMISSED** [review]

**Comment by gonzaponte** [comment]
I like this, but a second view made me realize that the `ndf` argument in `get_chi2_and_pvalue` is optional and set to 1. This shouldn't be like that, as the number of degrees of freedom is essential to the output. Also, at the risk to be pedantic again, can you mark the optional arguments in the docstring? Finally, we must add tests to ensure this new function is working properly and review those of the `fit` function which are probably very soft on the `chi2` output.

**Comment by jjgomezcadenas** [comment]
This review is stale, @gonzaponte, why? No advantage in letting open PRs which appear good enough to grow old. We are not running a winery here.

**Comment by gonzaponte** [comment]
> why?

Because we are still working on making it right once and for all. There are different interpretations of the parameters for scipy's `curve_fit` function. We want to handle all of them correctly. It seems stale, but we are pera-discussing it behind the scenes.

We have a working version which yields incorrect results only for the chi2 value and the errors of the parameters (which we are not paying too much attention to). We need to fix this, but it will not have a big impact on our lives.

File: invisible_cities/reco/corrections_test.py [hunk 4]
```diff
@@ -1,13 +1,14 @@
...
-from ..reco.corrections import LifetimeRCorrection
-from ..reco.corrections import LifetimeXYCorrection
+from ..core                import fit_functions as fitf
+from ..core.exceptions     import ParameterNotSet
+from ..reco.corrections    import Correction
+from ..reco.corrections    import Fcorrection
+from ..reco.corrections    import LifetimeCorrection
+from ..reco.corrections    import LifetimeRCorrection
+from ..reco.corrections    import LifetimeXYCorrection
+from ..icaro.hst_functions import poisson_sigma
```
Comment (gonzaponte) [inline invisible_cities/reco/corrections_test.py]: Please move `poisson_sigma` to `core_functions.py`. The `icaro` package will (probably) be removed in the near future.

File: invisible_cities/core/fit_functions.py [hunk 5]
```diff
@@ -137,8 +137,8 @@ def fit(func, x, y, seed=(), fit_range=None, **kwargs):
     fitx = fitf(x)
     nonzero = fitx != 0
 
-    sigma_r = kwargs.get("sigma", None)
-    if sigma_r is not None:
+    sigma_r = kwargs.get("sigma", 1)
+    if np.size(sigma_r) is not 1:
```
Comment (gonzaponte) [inline invisible_cities/core/fit_functions.py]: Mmm, I don't like this `size` check... How about setting the default to `np.ones` of the corresponding size and applying the `nonzero` selection over it?

File: invisible_cities/reco/corrections_test.py [hunk 6]
```diff
@@ -360,9 +361,10 @@ def test_corrections_1d(gauss_data_1d):
     mean = np.mean(Eevt)
     std  = np.std (Eevt)
 
-    y, x = np.histogram(Eevt, np.linspace(mean - 3 * std, mean + 3 * std, 100))
-    x    = x[:-1] + np.diff(x) * 0.5
-    f    = fitf.fit(fitf.gauss, x, y, (1e5, mean, std))
+    y, x  = np.histogram(Eevt, np.linspace(mean - 3 * std, mean + 3 * std, 100))
+    x     = x[:-1] + np.diff(x) * 0.5
```
Comment (gonzaponte) [inline invisible_cities/reco/corrections_test.py]: This is the same as the function `shift_to_bin_centers` in the `icaro` package. Can we move it also to `core` and use it here?

File: invisible_cities/reco/corrections_test.py [hunk 7]
```diff
@@ -380,5 +382,6 @@ def test_corrections_2d(gauss_data_2d):
 
     y, x = np.histogram(Eevt, np.linspace(mean - 3 * std, mean + 3 * std, 100))
     x    = x[:-1] + np.diff(x) * 0.5
-    f    = fitf.fit(fitf.gauss, x, y, (1e5, mean, std))
+    sigma = poisson_sigma(y)
+    f    = fitf.fit(fitf.gauss, x, y, (1e5, mean, std), sigma=sigma)
     assert f.chi2 < 3
```
Comment (gonzaponte) [inline invisible_cities/reco/corrections_test.py]: This `f.chi2 < 3` is maybe a bit soft. I think we can `flaky` these two tests.

**Review by gonzaponte — COMMENTED** [review]
Good work! This looks nice!

**Comment by jjgomezcadenas** [comment]
@gonzaponte I don't think that ICARO will be removed from IC!!!!

What we are going to do is to prototype in ICARO (repo), and the promote to IC/ICARO when we are happy enough, that means that PRS to IC/ICARO will continue but with the standard level of strictness (tests and such). The package will not disappear.

This said, poisson_sigma it's probably better in core

**Comment by gonzaponte** [comment]
@jjgomezcadenas Ok, then it is not obvious to me where these functions (those in `icaro/hst_functions.py` that are not for plotting) should go. I think we can postpone that discussion for another PR, do you agree?

**Comment by jjgomezcadenas** [comment]
Why not stat_tools.py also in IC/ICARO? Either now or in a different PR, your pick

**Comment by gonzaponte** [comment]
> Why not stat_tools.py also in IC/ICARO?

Agree. I was thinking about something like that too.

> Either now or in a different PR, your pick

I like to maintain PRs as self-contained as possible, so I would make a separate PR. But it can be performed by the same author as this one :)

**Comment by jjgomezcadenas** [comment]
What on earth happened here, @gonzaponte, @bpalmeiro? This is huger amount of work, we need to get it out of the way

**Comment by gonzaponte** [comment]
This is almost ready (I still need to review it a bit), however I am going to sit with @bpalmeiro to show him how to clean up the PR (it contains far too much noise). The time difference and the busy last weeks didn't help.

**Comment by jjgomezcadenas** [comment]
This has been opened for ever, @bpalmeiro, @gonzaponte . Are you serious or what?

**Comment by gonzaponte** [comment]
1. This is not urgent. There is nothing depending on this PR.
2. This has no conflict with TBR.
3. We will use this PR to practice some of the git-related operations and leave a clean history while reviewing with peras. However, @bpalmeiro is on vacation. I was going to do it by myself and redo it with him when he is back, but:
4. I don't have time AT ALL.
5. @bpalmeiro is back next week. We will begin the pera-review + cleanup then.
6. Again, there is nothing urgent here. Nothing in the data-processing chain depends on this. Please, keep calm, I am doing my best.

**Review by gonzaponte — APPROVED** [review]
This PR fixes a long-term issue with the definition of the chi2 and the p-value. The calculations have been encapsulated within a function, which is tested. The previous test suite for the rest of the code has also been fixed, improved and extended.

Great job!

**Pull request merged**